qiskit_nature/problems/sampling/protein_folding/bead_distances/distance_map_builder.py

```diff
@@ -0,0 +1,333 @@
+                delta_n2[lower_bead_ind][0][upper_bead_ind][0]
+            ).reduce()
+            delta_n3[lower_bead_ind][0][upper_bead_ind][0] = _fix_qubits(
+                delta_n3[lower_bead_ind][0][upper_bead_ind][0]
+            ).reduce()
+
+    return delta_n0, delta_n1, delta_n2, delta_n3
+
+
+def _init_distance_dict():
+    return collections.defaultdict(
+        lambda: collections.defaultdict(lambda: collections.defaultdict(dict))
+    )
+
+
+def _add_distances_side_chain(
+    peptide: Peptide,
+    delta_n0: DefaultDict[int, DefaultDict[int, DefaultDict[int, dict]]],
+    delta_n1: DefaultDict[int, DefaultDict[int, DefaultDict[int, dict]]],
+    delta_n2: DefaultDict[int, DefaultDict[int, DefaultDict[int, dict]]],
+    delta_n3: DefaultDict[int, DefaultDict[int, DefaultDict[int, dict]]],
+) -> Tuple[
+    DefaultDict[int, DefaultDict[int, DefaultDict[int, dict]]],
+    DefaultDict[int, DefaultDict[int, DefaultDict[int, dict]]],
+    DefaultDict[int, DefaultDict[int, DefaultDict[int, dict]]],
+    DefaultDict[int, DefaultDict[int, DefaultDict[int, dict]]],
+]:
+    """
+    Calculates distances between beads located on side chains and adds the contribution to the
+    distance calculated between beads (lower_bead_ind and upper_bead_ind) on the main chain. In
+    the absence
+    of side chains, this function returns a value of 0.
+
+    Args:
+        peptide: A Peptide object that includes all information about a protein.
+        delta_n0: Number of occurrences of axis 0 between beads.
+        delta_n1: Number of occurrences of axis 1 between beads.
+        delta_n2: Number of occurrences of axis 2 between beads.
+        delta_n3: Number of occurrences of axis 3 between beads.
+
+    Returns:
+        delta_n0, delta_n1, delta_n2, delta_n3: Updated tuple (with added side chain
+                                                contributions) that track the number
+                                                of occurrences of turns at axes 0,1,2,3.
+    """
+    main_chain_len = len(peptide.get_main_chain)
+    side_chain = peptide.get_side_chain_hot_vector()
+    for lower_bead_ind in range(1, main_chain_len):  # upper_bead_ind>lower_bead_ind
+        for upper_bead_ind in range(lower_bead_ind + 1, main_chain_len + 1):
+
+            if side_chain[upper_bead_ind - 1]:
+                try:
+                    # TODO generalize to side chains longer than 1
+                    indic_0, indic_1, indic_2, indic_3 = (
+                        peptide.get_main_chain[upper_bead_ind - 1]
+                        .side_chain[0]
+                        .get_indicator_functions()
+                    )
+                    delta_n0[lower_bead_ind][0][upper_bead_ind][1] = _fix_qubits(
+                        (
+                            delta_n0[lower_bead_ind][0][upper_bead_ind][0]
+                            + (-1) ** upper_bead_ind * indic_0
+                        )
+                    ).reduce()
```

Comment:
Again, this looks like a lot of duplicate code. Is there some way to improve this?